flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java

```diff
@@ -258,7 +258,7 @@ YarnLocalResourceDescriptor registerSingleLocalResource(
                     final java.nio.file.Path shipPath = file.toPath();
                     final java.nio.file.Path parentPath = shipPath.getParent();
                     Files.walkFileTree(
-                            shipPath,
+                            shipPath.toRealPath(),
```

Comment:
This will only solve the problem that `shipPath` is a symbolic link. What about if `shipPath` contains a symbolic link somewhere farther down? Wouldn't then something like

```
Files.walkFileTree(shipPath, EnumSet.of(FileVisitOption.FOLLOW_LINKS), Integer.MAX_VALUE, visitor);
```

be better suited?

Reply:
Sorry for my late reply. It's better for using `FileUtils.listFilesInDirectory` (flink-core) instead of `Files.walkFileTree()`.